File: app/src/main/java/com/dpis/module/SystemFontScaleToolBinder.java
package com.dpis.module;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.graphics.Typeface;
import android.net.Uri;
import android.provider.Settings;
import android.util.TypedValue;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.widget.AppCompatImageButton;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.card.MaterialCardView;
import com.google.android.material.slider.Slider;
import com.google.android.material.textview.MaterialTextView;

final class SystemFontScaleToolBinder {
    private static final float DISABLED_ACTION_ALPHA = 0.45f;
    private static final int PREVIEW_TITLE_SP = 18;
    private static final int PREVIEW_BODY_SP = 14;

    private final LocalizedActivity activity;
    private final View workspaceView;
    private final SystemFontScaleSettingsGateway settingsGateway;

    private MaterialCardView card;
    private View operationGroup;
    private View permissionOverlay;
    private View unavailableOverlay;
    private MaterialTextView badgeView;
    private MaterialTextView pendingValueView;
    private MaterialTextView previewTitleView;
    private MaterialTextView previewBodyView;
    private AppCompatImageButton applyButton;
    private AppCompatImageButton decrementButton;
    private AppCompatImageButton incrementButton;
    private Slider slider;
    private MaterialButton restoreButton;
    private boolean expanded;
    private boolean updatingSlider;
    private SystemFontScaleToolState state;

    SystemFontScaleToolBinder(LocalizedActivity activity, View workspaceView) {
        this(activity, workspaceView, new SystemFontScaleSettingsGateway());
    }

    SystemFontScaleToolBinder(LocalizedActivity activity,
                              View workspaceView,
                              SystemFontScaleSettingsGateway settingsGateway) {
        this.activity = activity;
        this.workspaceView = workspaceView;
        this.settingsGateway = settingsGateway;
    }

    void bind() {
        card = workspaceView.findViewById(R.id.system_font_scale_card);
        operationGroup = workspaceView.findViewById(R.id.system_font_scale_operation_group);
        permissionOverlay = workspaceView.findViewById(R.id.system_font_scale_permission_overlay);
        unavailableOverlay = workspaceView.findViewById(R.id.system_font_scale_unavailable_overlay);
        badgeView = workspaceView.findViewById(R.id.system_font_scale_badge);
        pendingValueView = workspaceView.findViewById(R.id.system_font_scale_pending_value);
        previewTitleView = workspaceView.findViewById(R.id.system_font_scale_preview_title);
        previewBodyView = workspaceView.findViewById(R.id.system_font_scale_preview_body);
        applyButton = workspaceView.findViewById(R.id.system_font_scale_apply_button);
        decrementButton = workspaceView.findViewById(R.id.system_font_scale_decrement_button);
        incrementButton = workspaceView.findViewById(R.id.system_font_scale_increment_button);
        slider = workspaceView.findViewById(R.id.system_font_scale_slider);
        restoreButton = workspaceView.findViewById(R.id.system_font_scale_restore_button);

        if (slider != null) {
            slider.setValueFrom(SystemFontScaleToolState.MIN_PERCENT);
            slider.setValueTo(SystemFontScaleToolState.MAX_PERCENT);
            slider.setStepSize(1f);
            slider.addOnChangeListener((slider, value, fromUser) -> {
                if (!fromUser || updatingSlider || state == null) {
                    return;
                }
                setPendingPercent(Math.round(value));
            });
        }

        TouchFeedbackBinder.bindPressHaptic(card);
        TouchFeedbackBinder.bindPressHaptic(applyButton);
        TouchFeedbackBinder.bindPressHaptic(decrementButton);
        TouchFeedbackBinder.bindPressHaptic(incrementButton);
        TouchFeedbackBinder.bindPressHaptic(restoreButton);

        if (card != null) {
            card.setOnClickListener(v -> {
                expanded = !expanded;
                render();
            });
        }
        if (applyButton != null) {
            applyButton.setOnClickListener(v -> applyPending());
        }
        if (decrementButton != null) {
            decrementButton.setOnClickListener(v -> {
                if (state != null && state.canDecrement()) {
                    setPendingPercent(state.pendingPercent - 1);
                }
            });
        }
        if (incrementButton != null) {
            incrementButton.setOnClickListener(v -> {
                if (state != null && state.canIncrement()) {
                    setPendingPercent(state.pendingPercent + 1);
                }
            });
        }
        if (restoreButton != null) {
            restoreButton.setOnClickListener(v -> restoreDefault());
        }
        TouchFeedbackBinder.bindPressHaptic(permissionOverlay);
        if (permissionOverlay != null) {
            permissionOverlay.setOnClickListener(v -> openWriteSettingsPermission());
        }

        refreshFromSystem();
    }

    void refreshFromSystem() {
        boolean canWrite = settingsGateway.canWrite(activity);
        Integer currentPercent = settingsGateway.readPercent(activity);
        boolean unavailable = currentPercent == null;
        int pendingPercent = state != null && state.userSelectedPending
                ? state.pendingPercent
                : SystemFontScaleToolState.initialPendingPercent(currentPercent);
        state = new SystemFontScaleToolState(
                canWrite,
                currentPercent,
                pendingPercent,
                state != null && state.userSelectedPending,
                unavailable);
        render();
    }

    void collapseAndRefreshFromSystem() {
        expanded = false;
        refreshFromSystem();
    }

    private void setPendingPercent(int percent) {
        if (state == null) {
            return;
        }
        state = new SystemFontScaleToolState(
                state.canWrite,
                state.currentPercent,
                SystemFontScaleToolState.clampPercent(percent),
                true,
                state.unavailable);
        render();
    }

    private void applyPending() {
        if (state == null || !state.canApply()) {
            return;
        }
        writeScale(state.pendingPercent);
    }

    private void restoreDefault() {
        if (state == null || !state.canRestore()) {
            return;
        }
        if (state.shouldRestorePendingOnly()) {
            setPendingPercent(SystemFontScaleToolState.DEFAULT_PERCENT);
            return;
        }
        writeScale(SystemFontScaleToolState.DEFAULT_PERCENT);
    }

    private void writeScale(int percent) {
        SystemFontScaleWriter.write(new SystemFontScaleWriter.Host() {
            @Override
            public boolean writePercent(int percent) {
                return settingsGateway.writePercent(activity, percent);
            }

            @Override
            public void onWriteSucceeded(int percent) {
                state = new SystemFontScaleToolState(
                        state.canWrite,
                        percent,
                        percent,
                        false,
                        false);
                refreshFromSystem();
            }

            @Override
            public void onWriteFailed() {
                showWriteFailed();
            }
        }, percent);
    }

    private void showWriteFailed() {
        Toast.makeText(
                activity,
                R.string.system_font_scale_write_failed,
                Toast.LENGTH_SHORT).show();
        refreshFromSystem();
    }

    private void openWriteSettingsPermission() {
        Intent intent = new Intent(Settings.ACTION_MANAGE_WRITE_SETTINGS);
        intent.setData(Uri.parse("package:" + activity.getPackageName()));
        activity.startActivity(intent);
    }

    @SuppressLint("SetTextI18n")
    private void render() {
        if (state == null) {
            return;
        }
        setVisible(operationGroup, expanded && (state.canWrite || state.unavailable));
        bindBadge();
        bindValues();
        bindControls();
        bindPreview();
        setVisible(permissionOverlay, expanded && !state.canWrite && !state.unavailable);
        setVisible(unavailableOverlay, expanded && state.unavailable);
    }

    private void bindBadge() {
        if (badgeView == null) {
            return;
        }
        int textRes;
        switch (state.badge()) {
            case UNAVAILABLE:
                textRes = R.string.system_font_scale_badge_unavailable;
                break;
            case PERMISSION_REQUIRED:
                textRes = R.string.system_font_scale_badge_permission_required;
                break;
            case OUT_OF_RANGE:
                textRes = R.string.system_font_scale_badge_out_of_range;
                break;
            case MODIFIED:
                textRes = R.string.system_font_scale_badge_modified;
                break;
            case NONE:
            default:
                textRes = 0;
                break;
        }
        if (textRes == 0) {
            badgeView.setVisibility(View.GONE);
            return;
        }
        badgeView.setVisibility(View.VISIBLE);
        badgeView.setText(textRes);
    }

    private void bindValues() {
        if (pendingValueView != null) {
            pendingValueView.setText(activity.getString(
                    R.string.system_font_scale_pending_value,
                    state.pendingPercent));
        }
    }

    private void bindControls() {
        boolean controlsEnabled = state.canWrite && !state.unavailable;
        if (slider != null) {
            updatingSlider = true;
            slider.setValue(SystemFontScaleToolState.clampPercent(state.pendingPercent));
            updatingSlider = false;
            slider.setEnabled(controlsEnabled);
        }
        bindIconButton(applyButton, state.canApply());
        bindIconButton(decrementButton, controlsEnabled && state.canDecrement());
        bindIconButton(incrementButton, controlsEnabled && state.canIncrement());
        if (restoreButton != null) {
            restoreButton.setEnabled(state.canRestore());
        }
    }

    private void bindIconButton(View button, boolean enabled) {
        if (button == null) {
            return;
        }
        button.setEnabled(enabled);
        button.setAlpha(enabled ? 1f : DISABLED_ACTION_ALPHA);
    }

    private void bindPreview() {
        float scale = state.pendingPercent / 100f;
        bindPreviewText(previewTitleView, PREVIEW_TITLE_SP, scale, true);
        bindPreviewText(previewBodyView, PREVIEW_BODY_SP, scale, false);
    }

    private void bindPreviewText(TextView view, int baseSp, float scale, boolean bold) {
        if (view == null) {
            return;
        }
        float density = activity.getResources().getDisplayMetrics().density;
        view.setTextSize(TypedValue.COMPLEX_UNIT_PX, baseSp * density * scale);
        view.setTypeface(Typeface.DEFAULT, bold ? Typeface.BOLD : Typeface.NORMAL);
    }

    private static void setVisible(View view, boolean visible) {
        if (view != null) {
            view.setVisibility(visible ? View.VISIBLE : View.GONE);
        }
    }
}

File: app/src/main/java/com/dpis/module/settings/presentation/SystemFontScaleToolBinder.kt
package com.dpis.module.settings.presentation

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Intent
import android.graphics.Rect
import android.graphics.Typeface
import android.net.Uri
import android.provider.Settings
import android.util.TypedValue
import android.view.View
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.widget.AppCompatImageButton
import com.dpis.module.R
import com.dpis.module.settings.SystemFontScaleSettingsGateway
import com.dpis.module.settings.SystemFontScaleToolState
import com.dpis.module.settings.SystemFontScaleWriter
import com.dpis.module.ui.WatchUiMode
import com.google.android.material.button.MaterialButton
import com.google.android.material.card.MaterialCardView
import com.google.android.material.slider.Slider
import com.google.android.material.textview.MaterialTextView

internal class SystemFontScaleToolBinder(
    private val activity: Activity,
    private val workspaceView: View,
    private val host: ToolsWorkspaceBinder.Host,
    private val settingsGateway: SystemFontScaleSettingsGateway = SystemFontScaleSettingsGateway(),
) {
    private var card: MaterialCardView? = null
    private var operationGroup: View? = null
    private var permissionOverlay: View? = null
    private var unavailableOverlay: View? = null
    private var badgeView: MaterialTextView? = null
    private var pendingValueView: MaterialTextView? = null
    private var previewTitleView: MaterialTextView? = null
    private var previewBodyView: MaterialTextView? = null
    private var applyButton: AppCompatImageButton? = null
    private var decrementButton: AppCompatImageButton? = null
    private var incrementButton: AppCompatImageButton? = null
    private var slider: Slider? = null
    private var restoreButton: MaterialButton? = null
    private var expanded = false
    private var updatingSlider = false
    private var state: SystemFontScaleToolState? = null

    fun bind() {
        card = workspaceView.findViewById(R.id.system_font_scale_card)
        operationGroup = workspaceView.findViewById(R.id.system_font_scale_operation_group)
        permissionOverlay = workspaceView.findViewById(R.id.system_font_scale_permission_overlay)
        unavailableOverlay = workspaceView.findViewById(R.id.system_font_scale_unavailable_overlay)
        badgeView = workspaceView.findViewById(R.id.system_font_scale_badge)
        pendingValueView = workspaceView.findViewById(R.id.system_font_scale_pending_value)
        previewTitleView = workspaceView.findViewById(R.id.system_font_scale_preview_title)
        previewBodyView = workspaceView.findViewById(R.id.system_font_scale_preview_body)
        applyButton = workspaceView.findViewById(R.id.system_font_scale_apply_button)
        decrementButton = workspaceView.findViewById(R.id.system_font_scale_decrement_button)
        incrementButton = workspaceView.findViewById(R.id.system_font_scale_increment_button)
        slider = workspaceView.findViewById(R.id.system_font_scale_slider)
        restoreButton = workspaceView.findViewById(R.id.system_font_scale_restore_button)

        slider?.let { sliderView ->
            sliderView.valueFrom = SystemFontScaleToolState.MIN_PERCENT.toFloat()
            sliderView.valueTo = SystemFontScaleToolState.MAX_PERCENT.toFloat()
            sliderView.stepSize = 1f
            sliderView.addOnChangeListener { _, value, fromUser ->
                if (!fromUser || updatingSlider || state == null) return@addOnChangeListener
                setPendingPercent(Math.round(value))
            }
        }

        host.bindPressHaptic(card)
        host.bindPressHaptic(applyButton)
        host.bindPressHaptic(decrementButton)
        host.bindPressHaptic(incrementButton)
        host.bindPressHaptic(restoreButton)

        card?.setOnClickListener {
            expanded = !expanded
            render()
            if (expanded) revealExpandedPanel()
        }
        applyButton?.setOnClickListener { applyPending() }
        decrementButton?.setOnClickListener {
            val current = state
            if (current != null && current.canDecrement()) {
                setPendingPercent(current.pendingPercent - 1)
            }
        }
        incrementButton?.setOnClickListener {
            val current = state
            if (current != null && current.canIncrement()) {
                setPendingPercent(current.pendingPercent + 1)
            }
        }
        restoreButton?.setOnClickListener { restoreDefault() }
        host.bindPressHaptic(permissionOverlay)
        permissionOverlay?.setOnClickListener { openWriteSettingsPermission() }

        refreshFromSystem()
    }

    fun refreshFromSystem() {
        val canWrite = settingsGateway.canWrite(activity)
        val currentPercent = settingsGateway.readPercent(activity)
        val unavailable = currentPercent == null
        val current = state
        val pendingPercent = if (current != null && current.userSelectedPending) {
            current.pendingPercent
        } else {
            SystemFontScaleToolState.initialPendingPercent(currentPercent)
        }
        state = SystemFontScaleToolState(
            canWrite,
            currentPercent,
            pendingPercent,
            current != null && current.userSelectedPending,
            unavailable,
        )
        render()
    }

    fun collapseAndRefreshFromSystem() {
        expanded = false
        refreshFromSystem()
    }

    private fun setPendingPercent(percent: Int) {
        val current = state ?: return
        state = SystemFontScaleToolState(
            current.canWrite,
            current.currentPercent,
            SystemFontScaleToolState.clampPercent(percent),
            true,
            current.unavailable,
        )
        render()
    }

    private fun applyPending() {
        val current = state
        if (current == null || !current.canApply()) return
        writeScale(current.pendingPercent)
    }

    private fun restoreDefault() {
        val current = state
        if (current == null || !current.canRestore()) return
        if (current.shouldRestorePendingOnly()) {
            setPendingPercent(SystemFontScaleToolState.DEFAULT_PERCENT)
            return
        }
        writeScale(SystemFontScaleToolState.DEFAULT_PERCENT)
    }

    private fun writeScale(percent: Int) {
        SystemFontScaleWriter.write(object : SystemFontScaleWriter.Host {
            override fun writePercent(value: Int): Boolean =
                settingsGateway.writePercent(activity, value)

            override fun onWriteSucceeded(value: Int) {
                val current = state
                state = SystemFontScaleToolState(
                    current?.canWrite ?: false,
                    value,
                    value,
                    false,
                    false,
                )
                refreshFromSystem()
            }

            override fun onWriteFailed() {
                showWriteFailed()
            }
        }, percent)
    }

    private fun showWriteFailed() {
        Toast.makeText(
            activity,
            R.string.system_font_scale_write_failed,
            Toast.LENGTH_SHORT,
        ).show()
        refreshFromSystem()
    }

    private fun openWriteSettingsPermission() {
        val intent = Intent(Settings.ACTION_MANAGE_WRITE_SETTINGS)
        intent.data = Uri.parse("package:" + activity.packageName)
        activity.startActivity(intent)
    }

    @SuppressLint("SetTextI18n")
    private fun render() {
        val current = state ?: return
        setVisible(operationGroup, expanded && (current.canWrite || current.unavailable))
        bindBadge()
        bindValues()
        bindControls()
        bindPreview()
        setVisible(permissionOverlay, expanded && !current.canWrite && !current.unavailable)
        setVisible(unavailableOverlay, expanded && current.unavailable)
    }

    private fun bindBadge() {
        val badge = badgeView ?: return
        val current = state ?: return
        val textRes = when (current.badge()) {
            SystemFontScaleToolState.Badge.UNAVAILABLE ->
                R.string.system_font_scale_badge_unavailable
            SystemFontScaleToolState.Badge.PERMISSION_REQUIRED -> {
                if (WatchUiMode.shouldUseCompactUi(activity)) {
                    badge.visibility = View.GONE
                    return
                }
                R.string.system_font_scale_badge_permission_required
            }
            SystemFontScaleToolState.Badge.OUT_OF_RANGE ->
                R.string.system_font_scale_badge_out_of_range
            SystemFontScaleToolState.Badge.MODIFIED ->
                R.string.system_font_scale_badge_modified
            SystemFontScaleToolState.Badge.NONE -> 0
        }
        if (textRes == 0) {
            badge.visibility = View.GONE
            return
        }
        badge.visibility = View.VISIBLE
        badge.setText(textRes)
    }

    private fun bindValues() {
        val current = state ?: return
        pendingValueView?.text = activity.getString(
            R.string.system_font_scale_pending_value,
            current.pendingPercent,
        )
    }

    private fun bindControls() {
        val current = state ?: return
        val controlsEnabled = current.canWrite && !current.unavailable
        slider?.let { sliderView ->
            updatingSlider = true
            sliderView.value = SystemFontScaleToolState.clampPercent(current.pendingPercent).toFloat()
            updatingSlider = false
            sliderView.isEnabled = controlsEnabled
        }
        bindIconButton(applyButton, current.canApply())
        bindIconButton(decrementButton, controlsEnabled && current.canDecrement())
        bindIconButton(incrementButton, controlsEnabled && current.canIncrement())
        restoreButton?.isEnabled = current.canRestore()
    }

    private fun bindIconButton(button: View?, enabled: Boolean) {
        if (button == null) return
        button.isEnabled = enabled
        button.alpha = if (enabled) 1f else DISABLED_ACTION_ALPHA
    }

    private fun bindPreview() {
        val current = state ?: return
        val scale = current.pendingPercent / 100f
        bindPreviewText(previewTitleView, PREVIEW_TITLE_SP, scale, true)
        bindPreviewText(previewBodyView, PREVIEW_BODY_SP, scale, false)
    }

    private fun bindPreviewText(view: TextView?, baseSp: Int, scale: Float, bold: Boolean) {
        if (view == null) return
        val density = activity.resources.displayMetrics.density
        view.setTextSize(TypedValue.COMPLEX_UNIT_PX, baseSp * density * scale)
        view.setTypeface(Typeface.DEFAULT, if (bold) Typeface.BOLD else Typeface.NORMAL)
    }

    private fun revealExpandedPanel() {
        val target = visibleExpandedPanel() ?: return
        target.post {
            val bounds = Rect(0, 0, target.width, target.height)
            target.requestRectangleOnScreen(bounds, true)
        }
    }

    private fun visibleExpandedPanel(): View? {
        val permission = permissionOverlay
        if (permission != null && permission.visibility == View.VISIBLE) return permission
        val operations = operationGroup
        if (operations != null && operations.visibility == View.VISIBLE) return operations
        return null
    }

    companion object {
        private const val DISABLED_ACTION_ALPHA = 0.45f
        private const val PREVIEW_TITLE_SP = 18
        private const val PREVIEW_BODY_SP = 14

        private fun setVisible(view: View?, visible: Boolean) {
            if (view != null) {
                view.visibility = if (visible) View.VISIBLE else View.GONE
            }
        }
    }
}
